sql/src/main/java/io/crate/expression/symbol/Symbol.java

```diff
@@ -58,7 +59,9 @@ public final Symbol cast(DataType<?> targetType) {
      * @return An instance of {@link Function} which casts this symbol.
      */
     public Symbol cast(DataType<?> targetType, boolean tryCast) {
-        if (targetType.equals(valueType())) {
+        // TODO implement is convertible for objects
```

Comment:
Added this condition just for now. do you think we have to have a proper `equal` and `isConvertible` for object types, by proper i mean taking into account inner types and considering cases with different fields, undefined types, etc. I am not sure how straightforward that can be.

Reply:
Can you elaborate on the `isConvertible` relationship to `equals`?

I think we should override equals on `ObjectType` to take inner-types into consideration.

Reply:
because `isConvertableTo` uses `equals`. We can have cases where we check whether some literal of an object type that missing the inner types info checked whether it can be converted to target object type with some inner types. It can also be other way around, where the source type has inner types and the target type missing them. I'd say that it is still ok to covert object to object if one missing the inner type info completely, but implementing the `equal` in this way seems to be not that correct. Therefore, i was talking about special handling of object types in `isConvertible`, but not sure :/